// @embroider/core/src/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.optionsWithDefaults = optionsWithDefaults;
function optionsWithDefaults(options) {
    let staticComponents, staticHelpers, staticModifiers;
    // staticInvokables always wins when configured
    if (typeof (options === null || options === void 0 ? void 0 : options.staticInvokables) !== 'undefined') {
        if (typeof options.staticComponents !== 'undefined' ||
            typeof options.staticHelpers !== 'undefined' ||
            typeof options.staticModifiers !== 'undefined') {
            throw new Error('You cannot set `staticHelpers`, `staticComponents`, or `staticModifiers` if you have set `staticInvokables`. Delete these configs to continue.');
        }
        staticComponents = staticHelpers = staticModifiers = options.staticInvokables;
    }
    else {
        if (typeof (options === null || options === void 0 ? void 0 : options.staticComponents) !== 'undefined') {
            console.error(`Setting 'staticComponents' is deprecated. Use 'staticInvokables' instead`);
            staticComponents = options.staticComponents;
        }
        if (typeof (options === null || options === void 0 ? void 0 : options.staticHelpers) !== 'undefined') {
            console.error(`Setting 'staticHelpers' is deprecated. Use 'staticInvokables' instead`);
            staticHelpers = options.staticHelpers;
        }
        if (typeof (options === null || options === void 0 ? void 0 : options.staticModifiers) !== 'undefined') {
            console.error(`Setting 'staticModifiers' is deprecated. Use 'staticInvokables' instead`);
            staticModifiers = options.staticModifiers;
        }
    }
    let defaults = {
        splitAtRoutes: [],
        staticAppPaths: [],
        skipBabel: [],
        pluginHints: [],
        staticHelpers: staticHelpers !== null && staticHelpers !== void 0 ? staticHelpers : false,
        staticModifiers: staticModifiers !== null && staticModifiers !== void 0 ? staticModifiers : false,
        staticComponents: staticComponents !== null && staticComponents !== void 0 ? staticComponents : false,
        amdCompatibility: 'cjs',
    };
    if (options) {
        return Object.assign(defaults, options);
    }
    return defaults;
}
