// @embroider/compat/src/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.recommendedOptions = void 0;
exports.optionsWithDefaults = optionsWithDefaults;
const core_1 = require("@embroider/core");
const defaults = Object.assign((0, core_1.optionsWithDefaults)(), {
    staticAddonTrees: false,
    staticAddonTestSupportTrees: false,
    staticEmberSource: false,
    compatAdapters: new Map(),
    extraPublicTrees: [],
    workspaceDir: null,
    packageRules: [],
    allowUnsafeDynamicComponents: false,
});
function optionsWithDefaults(options) {
    var _a, _b, _c, _d, _e, _f;
    if (!(options === null || options === void 0 ? void 0 : options.staticEmberSource)) {
        console.log(`The setting 'staticEmberSource' will default to true in the next version of Embroider and can't be turned off. To prepare for this you should set 'staticEmberSource: true' in your Embroider config.`);
    }
    if (!(options === null || options === void 0 ? void 0 : options.staticAddonTrees)) {
        console.log(`The setting 'staticAddonTrees' will default to true in the next version of Embroider and can't be turned off. To prepare for this you should set 'staticAddonTrees: true' in your Embroider config.`);
    }
    if (!(options === null || options === void 0 ? void 0 : options.staticAddonTestSupportTrees)) {
        console.log(`The setting 'staticAddonTestSupportTrees' will default to true in the next version of Embroider and can't be turned off. To prepare for this you should set 'staticAddonTestSupportTrees: true' in your Embroider config.`);
    }
    return Object.assign({}, defaults, {
        ...options,
        staticComponents: (_b = (_a = options === null || options === void 0 ? void 0 : options.staticInvokables) !== null && _a !== void 0 ? _a : options === null || options === void 0 ? void 0 : options.staticComponents) !== null && _b !== void 0 ? _b : defaults.staticComponents,
        staticHelpers: (_d = (_c = options === null || options === void 0 ? void 0 : options.staticInvokables) !== null && _c !== void 0 ? _c : options === null || options === void 0 ? void 0 : options.staticHelpers) !== null && _d !== void 0 ? _d : defaults.staticHelpers,
        staticModifiers: (_f = (_e = options === null || options === void 0 ? void 0 : options.staticInvokables) !== null && _e !== void 0 ? _e : options === null || options === void 0 ? void 0 : options.staticModifiers) !== null && _f !== void 0 ? _f : defaults.staticModifiers,
    });
}
// These are recommended configurations for addons to test themselves under. By
// keeping them here, it's easier to do ecosystem-wide compatibility testing.
// See the `@embroider/test-setup` package which can help consume these to test
// them in CI.
exports.recommendedOptions = Object.freeze({
    safe: Object.freeze({}),
    optimized: Object.freeze({
        staticAddonTrees: true,
        staticAddonTestSupportTrees: true,
        staticEmberSource: true,
        allowUnsafeDynamicComponents: false,
        staticInvokables: true,
    }),
});
