// @embroider/compat/src/compat-adapters/ember-fetch.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_extra_1 = require("fs-extra");
const add_to_tree_1 = __importDefault(require("../add-to-tree"));
const v1_addon_1 = __importDefault(require("../v1-addon"));
const path_1 = require("path");
const fs_1 = require("fs");
class default_1 extends v1_addon_1.default {
    get packageMeta() {
        let result = super.packageMeta;
        result['renamed-modules'] = {
            fetch: 'ember-fetch/index.js',
        };
        return result;
    }
    get v2Tree() {
        // ember-fetch attempts to emits an AMD define into vendor.js. We instead
        // adapt it into an ES module, so it can continue accessing ember even once
        // ember is not in AMD.
        return new add_to_tree_1.default(super.v2Tree, (outputPath) => {
            let vendorFile = (0, path_1.resolve)(outputPath, 'vendor/ember-fetch.js');
            let src = (0, fs_extra_1.readFileSync)(vendorFile, 'utf8');
            // the addon has already done "app.import('vendor/ember-fetch.js')", which
            // pushes it onto a list down inside the guts of ember-cli. If we delete
            // the file, we'll get a crash when ember-cli goes looking for it.
            // Instead, we just make it empty.
            (0, fs_1.writeFileSync)(vendorFile, '');
            (0, fs_1.writeFileSync)((0, path_1.resolve)(outputPath, 'index.js'), `import Ember from 'ember';
const outputs = {};

function define(name, deps, fn) {
  if (name !== 'fetch') {
    throw new Error('bug: we were attempting to capture ember-fetch');
  }
  if (deps[0] !== 'exports' && deps[1] !== 'ember') {
    throw new Error('bug: we unexpected deps while capturing ember-fetch');
  }
  fn(outputs, Ember);
}
${src}
export default outputs.default;
export const Headers = outputs.Headers;
export const Request = outputs.Request;
export const Response = outputs.Response;
export const AbortController = outputs.AbortController;
export const AbortSignal = outputs.AbortSignal;
export const fetch = outputs.fetch;
`);
        });
    }
}
exports.default = default_1;
